geode-core/src/test/java/org/apache/geode/internal/cache/versions/AbstractVersionTagTestBase.java

```diff
@@ -31,6 +64,122 @@ public void setup() {
     this.vt = createVersionTag();
   }
 
+  int getRandomUnusedInt() {
+    int unusedInt;
+    do {
+      unusedInt = random.nextInt(60000);
+    } while (usedInts.contains(unusedInt));
+    return unusedInt;
+  }
+
+  @Test
+  public void testConcurrentCanonicalizationOfIDsAndSerialization() throws IOException {
+    VersionTag spy = spy(vt);
+    DataOutput dataOutput = mock(DataOutput.class);
+    spy.setMemberID(createMemberID());
+    spy.setPreviousMemberID(createMemberID());
+    final short[] flags = {0};
+
+    Answer myAnswer = new Answer() {
+      boolean firstInvocation = true;
+
+      @Override
+      public Object answer(InvocationOnMock invocation) throws Throwable {
+        if (firstInvocation) {
+          // save the argument - it's the "flags" int that we'll want to verify
+          flags[0] = (short) (((Integer) invocation.getArgument(0)).intValue() & 0xFFFF);
+          firstInvocation = false;
+          // canonicalize the member IDs. Once flags have been written the tag shouldn't examine
+          // previousMemberID again to see if it's the same as the memberID.
+          spy.setPreviousMemberID(spy.getMemberID());
+        }
+        return null;
+      }
+    };
+    doAnswer(myAnswer).when(dataOutput).writeShort(any(Integer.class));
+    spy.toData(dataOutput, true);
+    // verify that we only wrote the
+    verify(spy, times(2)).writeMember(isA(VersionSource.class), isA(DataOutput.class));
```

Comment:
From the method name `testConcurrent…` I get the impression that this test is meant to verify that the bug (GEODE-7219) is fixed. I went to what is now line 385 of `VersionTag.java` and reverted what seems to be the primary fix, and then changed this spy to `times(1)` and this test passes (I expected it to fail). I ran `VMVersionTagTest` and `DiskVersionTagTest`—the two subclasses, and they both pass.

Reply:
That bit is complicated by the other change I made - to use Objects.equals() instead of "==" for equality checks.

Reply:
If you revert all of the changes to VersionTag.toData I think you'll find that the test fails with the method only writing one member ID.  That's the error that we're actually seeing cause the BufferUnderflowExceptions.  You'll also see the other new test "testSerializationWritesNoMemberID" failing because the old logic was writing out the previousMemberID even though it was being told explicitly to _not_ write any member IDs.

Reply:
With changes reverted I see one test failing: `testBufferUnderflowFromOldVersionIsIgnored()`. It fails with `BufferUnderflowException` ✓

Reply:
All three new tests should be failing if you revert all changes in toData

Reply:
confirmed

Reply:
bill told me in a pm that those other two tests also failed for him